File: 42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int sz = height.size();
        int lft_mx = 0, rgt_mx = 0;
        int ans = 0;
        int lft = 0, rgt = sz - 1;
        while( lft < rgt ) { 

            if(height[lft] < height[rgt] ) { 
                height[lft] >= lft_mx ? ( lft_mx = height[lft] ) : ans += ( lft_mx - height[lft]);
                lft++;

            } else { 
                height[rgt] >= rgt_mx ? ( rgt_mx = height[rgt] ) : ans += ( rgt_mx - height[rgt]);
                rgt--;

            }

        }
        return ans;
        
        
    }
};

/* 


*/ 

File: python/Lakshay/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) <= 2:
            return 0
        
        ans = 0
        
        sz = len(height)
        left_mx = [0] * sz
        right_mx = [0] * sz
        left_mx[0] = height[0]
        for i in range( 1 , sz ):
            left_mx[i] = max( left_mx[i-1], height[i])
        
        right_mx[sz-1] = height[sz - 1]
        for i in range(sz - 2, -1, -1 ):
            right_mx[i] = max( right_mx[i+1], height[i])
        
        for i in range(i, sz - 1):
            ans += min( left_mx[i], right_mx[i] ) - height[i];
        
        return ans
            
